File: Freedom/微信/微信/ChatDetailViewController/WXCGroupDetailViewController.m
//  FreedomGroupDetailViewController.m
//  Freedom
// Created by Super
#import "WXCGroupDetailViewController.h"
#import "WXUserGroupCell.h"
#import <BlocksKit/BlocksKit+UIKit.h>
#import "WXChatViewController.h"
#import "WXMessageManager.h"
#import "WXFriendDetailViewController.h"
#import "WXGroupQRCodeViewController.h"
#import "WXChatFileViewController.h"
#import "WXBgSettingViewController.h"
@interface WXCGroupDetailViewController () <WechatUserGroupCellDelegate>
@property (nonatomic, strong) WXMessageManager *helper;
@end
@implementation WXCGroupDetailViewController
- (void)viewDidLoad {
    [super viewDidLoad];
    [self.navigationItem setTitle:@"聊天详情"];
    
    self.helper = [WXMessageManager sharedInstance];
    self.data = [self.helper chatDetailDataByGroupInfo:self.group];
    
    [self.tableView registerClass:[WXUserGroupCell class] forCellReuseIdentifier:@"TLUserGroupCell"];
}
#pragma mark - Delegate -
//MARK: UITableViewDataSource
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.section == 0 && indexPath.row == 0) {
        WXUserGroupCell *cell = [tableView dequeueReusableCellWithIdentifier:@"TLUserGroupCell"];
        [cell setUsers:self.group.users];
        [cell setDelegate:self];
        return cell;
    }
    return [super tableView:tableView cellForRowAtIndexPath:indexPath];
}
//MARK: UITableViewDelegate
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    WXSettingItem *item = [self.data[indexPath.section] objectAtIndex:indexPath.row];
    if ([item.title isEqualToString:@"群二维码"]) {
        WXGroupQRCodeViewController *gorupQRCodeVC = [[WXGroupQRCodeViewController alloc] init];
        [gorupQRCodeVC setGroup:self.group];
        [self setHidesBottomBarWhenPushed:YES];
        [self.navigationController pushViewController:gorupQRCodeVC animated:YES];
    }else if ([item.title isEqualToString:@"设置当前聊天背景"]) {
        WXBgSettingViewController *chatBGSettingVC = [[WXBgSettingViewController alloc] init];
        [chatBGSettingVC setPartnerID:self.group.groupID];
        [self setHidesBottomBarWhenPushed:YES];
        [self.navigationController pushViewController:chatBGSettingVC animated:YES];
    }else if ([item.title isEqualToString:@"聊天文件"]) {
        WXChatFileViewController *chatFileVC = [[WXChatFileViewController alloc] init];
        [chatFileVC setPartnerID:self.group.groupID];
        [self setHidesBottomBarWhenPushed:YES];
        [self.navigationController pushViewController:chatFileVC animated:YES];
    }else if ([item.title isEqualToString:@"清空聊天记录"]) {
        [self showAlerWithtitle:nil message:nil style:UIAlertControllerStyleActionSheet ac1:^UIAlertAction *{
            return [UIAlertAction actionWithTitle:@"清空聊天记录" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                BOOL ok = [[WXMessageManager sharedInstance] deleteMessagesByPartnerID:self.group.groupID];
                if (!ok) {
                    [SVProgressHUD showErrorWithStatus:@"清空讨论组聊天记录失败"];
                }else{
                    [[WXChatViewController sharedChatVC] resetChatVC];
                }
            }];
        } ac2:^UIAlertAction *{
            return [UIAlertAction actionWithTitle:@"取消" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
            }];
        } ac3:nil completion:nil];
    }
    [self.tableView deselectRowAtIndexPath:indexPath animated:NO];
}
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.section == 0 && indexPath.row == 0) {
        NSUInteger count = self.group.count;
        return ((count + 1) / 4 + ((((count + 1) % 4) == 0) ? 0 : 1)) * 90 + 15;
    }
    return [super tableView:tableView heightForRowAtIndexPath:indexPath];
}
//MARK: TLUserGroupCellDelegate
- (void)userGroupCellDidSelectUser:(WXUser *)user{
    WXFriendDetailViewController *detailVC = [[WXFriendDetailViewController alloc] init];
    [detailVC setUser:user];
    [self setHidesBottomBarWhenPushed:YES];
    [self.navigationController pushViewController:detailVC animated:YES];
}
- (void)userGroupCellAddUserButtonDown{
    [SVProgressHUD showInfoWithStatus:@"添加讨论组成员"];
}
@end

File: Freedom/微信/微信/ChatDetailViewController/WXCGroupDetailViewController.swift
//
//  WXCGroupDetailViewController.swift
//  Freedom

import Foundation
class WXCGroupDetailViewController: WXSettingViewController, WechatUserGroupCellDelegate {
    var group: WXGroup = WXGroup()
    var helper = WXMessageManager.shared
    override func viewDidLoad() {
        super.viewDidLoad()
        navigationItem.title = "聊天详情"
        data = helper.chatDetailData(byGroupInfo: group)
        tableView.register(WXUserGroupCell.self, forCellReuseIdentifier: "TLUserGroupCell")
    }
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if indexPath.section == 0 && indexPath.row == 0 {
            let cell = tableView.dequeueReusableCell(withIdentifier: "TLUserGroupCell") as! WXUserGroupCell
            cell.users = group.users
            cell.delegate = self
            return cell
        }
        return super.tableView(tableView, cellForRowAt: indexPath)
    }
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let item = data[indexPath.section].items[indexPath.row]
        if (item.title == "群二维码") {
            let gorupQRCodeVC = WXGroupQRCodeViewController()
            gorupQRCodeVC.group = group
            hidesBottomBarWhenPushed = true
            navigationController?.pushViewController(gorupQRCodeVC, animated: true)
        } else if (item.title == "设置当前聊天背景") {
            let chatBGSettingVC = WXBgSettingViewController()
            chatBGSettingVC.partnerID = group.groupID
            hidesBottomBarWhenPushed = true
            navigationController?.pushViewController(chatBGSettingVC, animated: true)
        } else if (item.title == "聊天文件") {
            let chatFileVC = WXChatFileViewController()
            chatFileVC.partnerID = group.groupID
            hidesBottomBarWhenPushed = true
            navigationController?.pushViewController(chatFileVC, animated: true)
        } else if (item.title == "清空聊天记录") {
            let alertVC = UIAlertController("", "", T1: "清空聊天记录", T2: "取消", confirm1: {
                let ok = WXMessageManager.shared.deleteMessages(byPartnerID: self.group.groupID)
                WXChatViewController.shared.resetChatVC()
            }, confirm2: {

            })
            self.present(alertVC, animated: true, completion: nil)
        }
        self.tableView.deselectRow(at: indexPath, animated: false)
    }
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        if indexPath.section == 0 && indexPath.row == 0 {
            let count = group.count
            return CGFloat(((count + 1) / 4 + ((((count + 1) % 4) == 0) ? 0 : 1)) * 90 + 15)
        }
        return super.tableView(tableView, heightForRowAt: indexPath)
    }
    func userGroupCellDidSelect(_ user: WXUser) {
        let detailVC = WXFriendDetailViewController()
        detailVC.user = user
        hidesBottomBarWhenPushed = true
        navigationController?.pushViewController(detailVC, animated: true)
    }
    func userGroupCellAddUserButtonDown() {
        noticeInfo("添加讨论组成员")
    }
}
